shell/renderer/api/atom_api_context_bridge.cc

```diff
@@ -0,0 +1,408 @@
+    {
+      v8::Context::Scope destination_scope(destination);
+      v8::Local<v8::Value> proxy_func = mate::ConvertToV8(
+          destination->GetIsolate(),
+          base::BindRepeating(&ProxyFunctionWrapper, store, func_id));
+      FunctionLifeMonitor::BindTo(destination->GetIsolate(),
+                                  v8::Local<v8::Object>::Cast(proxy_func),
+                                  store, func_id);
+      return proxy_func;
+    }
+  }
+
+  // Proxy promises as they have a safe and guarunteed memory lifecycle (unlike
+  // functions)
+  if (value->IsPromise()) {
+    v8::Context::Scope destination_scope(destination);
+
+    auto v8_promise = v8::Local<v8::Promise>::Cast(value);
+    auto* promise =
+        new util::Promise<v8::Local<v8::Value>>(destination->GetIsolate());
+    v8::Local<v8::Promise> handle = promise->GetHandle();
+
+    auto then_cb = base::BindOnce(
+        [](util::Promise<v8::Local<v8::Value>>* promise, v8::Isolate* isolate,
+           v8::Global<v8::Context> source, v8::Global<v8::Context> destination,
+           RenderFramePersistenceStore* store, v8::Local<v8::Value> result) {
+          promise->Resolve(PassValueToOtherContext(
+              source.Get(isolate), destination.Get(isolate), result, store));
+          delete promise;
+        },
+        promise, destination->GetIsolate(),
+        v8::Global<v8::Context>(source->GetIsolate(), source),
+        v8::Global<v8::Context>(destination->GetIsolate(), destination), store);
+    auto catch_cb = base::BindOnce(
+        [](util::Promise<v8::Local<v8::Value>>* promise, v8::Isolate* isolate,
+           v8::Global<v8::Context> source, v8::Global<v8::Context> destination,
+           RenderFramePersistenceStore* store, v8::Local<v8::Value> result) {
+          promise->Reject(PassValueToOtherContext(
+              source.Get(isolate), destination.Get(isolate), result, store));
+          delete promise;
+        },
+        promise, destination->GetIsolate(),
+        v8::Global<v8::Context>(source->GetIsolate(), source),
+        v8::Global<v8::Context>(destination->GetIsolate(), destination), store);
+
+    ignore_result(v8_promise->Then(
+        source,
+        v8::Local<v8::Function>::Cast(
+            mate::ConvertToV8(destination->GetIsolate(), then_cb)),
+        v8::Local<v8::Function>::Cast(
+            mate::ConvertToV8(destination->GetIsolate(), catch_cb))));
+
+    return handle;
+  }
+
+  // Errors aren't serializable currently, we need to pull the message out and
+  // re-construct in the destination context
+  if (value->IsNativeError()) {
+    v8::Context::Scope destination_context_scope(destination);
+    return v8::Exception::Error(
+        v8::Exception::CreateMessage(destination->GetIsolate(), value)->Get());
+  }
+
+  // Manually go through the array and pass each value individually into a new
+  // array so that deep functions get proxied or arrays of promises.
```

Comment:
I'm not sure I understand this sentence ("deep functions get proxied or arrays of promises"...?)

Could you rephrase?

Reply:
Rephrased